src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -180,12 +180,16 @@ def make_requirement_from_install_req(self, ireq):
             # TODO: Get name and version from ireq, if possible?
             #       Specifically, this might be needed in "name @ URL"
             #       syntax - need to check where that syntax is handled.
-            cand = self._make_candidate_from_link(
+            candidate = self._make_candidate_from_link(
                 ireq.link, extras=set(ireq.extras), parent=ireq,
             )
-            return ExplicitRequirement(cand)
+            return self.make_requirement_from_candidate(candidate)
```

Comment:
Not really related, but this might read better if we flip the conditionals (`if not ireq.link:`)

Reply:
Didn't I see exactly that change in a different PR I was reviewing a few minutes ago? There may be merge conflicts in our future... But yes, I agree switching the order reads better.

Reply:
Yep, #8066

Reply:
Seems like my coding style is fairly consistent 😆

Reply:
I wanted to make the same change, but resisted the urge in favor of keeping the PR smaller. :)

Having this come in via #8066 works well for me.